File: tests/test_input_sources.py
from __future__ import annotations

import io
import subprocess
import sys

from lcpush import clipboard, editor


def test_read_stdin_until_eof_sentinel():
    stream = io.StringIO("line one\nline two\nEOF\nignored\n")
    assert editor.read_stdin(stream) == "line one\nline two\n"


def test_read_stdin_until_end_of_stream():
    assert editor.read_stdin(io.StringIO("a\nb\n")) == "a\nb\n"


def test_editor_command_prefers_visual(monkeypatch):
    monkeypatch.setenv("VISUAL", "code --wait")
    monkeypatch.setenv("EDITOR", "nano")
    assert editor.editor_command() == ["code", "--wait"]


def test_editor_command_falls_back_to_vi(monkeypatch):
    monkeypatch.delenv("VISUAL", raising=False)
    monkeypatch.delenv("EDITOR", raising=False)
    assert editor.editor_command() == ["vi"]


def test_strip_header_removes_the_instruction_line():
    body = editor.SOLUTION_HEADER + "print(1)\n"
    assert editor.strip_header(body, editor.SOLUTION_HEADER).strip() == "print(1)"


def test_open_editor_returns_none_when_unchanged(monkeypatch):
    monkeypatch.setenv("EDITOR", "true")
    monkeypatch.setattr(subprocess, "run", lambda *a, **k: None)
    assert editor.open_editor() is None


def test_open_editor_returns_written_content(monkeypatch):
    monkeypatch.setenv("EDITOR", "fake")

    def fake_run(command, **_kwargs):
        path = command[-1]
        with open(path, "w", encoding="utf-8") as handle:
            handle.write(editor.SOLUTION_HEADER + "class Solution: pass\n")

    monkeypatch.setattr(subprocess, "run", fake_run)
    assert editor.open_editor().strip() == "class Solution: pass"


def test_open_editor_returns_none_for_empty_buffer(monkeypatch):
    monkeypatch.setenv("EDITOR", "fake")

    def fake_run(command, **_kwargs):
        with open(command[-1], "w", encoding="utf-8") as handle:
            handle.write("\n\n")

    monkeypatch.setattr(subprocess, "run", fake_run)
    assert editor.open_editor() is None


def test_clipboard_command_on_macos(monkeypatch):
    monkeypatch.setattr(sys, "platform", "darwin")
    monkeypatch.setattr(clipboard.shutil, "which", lambda name: "/usr/bin/pbpaste")
    assert clipboard.clipboard_command() == ["pbpaste"]


def test_clipboard_command_missing_tool(monkeypatch):
    monkeypatch.setattr(sys, "platform", "linux")
    monkeypatch.delenv("WAYLAND_DISPLAY", raising=False)
    monkeypatch.setattr(clipboard.shutil, "which", lambda _name: None)
    assert clipboard.clipboard_command() is None
    assert not clipboard.available()
    assert clipboard.read() is None


def test_clipboard_prefers_wayland(monkeypatch):
    monkeypatch.setattr(sys, "platform", "linux")
    monkeypatch.setenv("WAYLAND_DISPLAY", "wayland-0")
    monkeypatch.setattr(clipboard.shutil, "which", lambda name: f"/usr/bin/{name}")
    assert clipboard.clipboard_command()[0] == "wl-paste"


def test_clipboard_read_returns_none_when_whitespace(monkeypatch):
    monkeypatch.setattr(clipboard, "clipboard_command", lambda: ["echo"])

    class Result:
        returncode = 0
        stdout = "   \n"

    monkeypatch.setattr(clipboard.subprocess, "run", lambda *a, **k: Result())
    assert clipboard.read() is None


def test_clipboard_read_returns_content(monkeypatch):
    monkeypatch.setattr(clipboard, "clipboard_command", lambda: ["echo"])

    class Result:
        returncode = 0
        stdout = "class Solution: pass\n"

    monkeypatch.setattr(clipboard.subprocess, "run", lambda *a, **k: Result())
    assert clipboard.read() == "class Solution: pass\n"


def test_clipboard_read_survives_a_failing_tool(monkeypatch):
    monkeypatch.setattr(clipboard, "clipboard_command", lambda: ["nope"])

    def boom(*_args, **_kwargs):
        raise OSError("not found")

    monkeypatch.setattr(clipboard.subprocess, "run", boom)
    assert clipboard.read() is None

File: tests/test_input_sources.cpp
#include <catch2/catch_test_macros.hpp>

#include <fstream>
#include <sstream>

#include "helpers/fake_subprocess.hpp"
#include "helpers/fixtures.hpp"
#include "lcpush/clipboard.hpp"
#include "lcpush/editor.hpp"

using namespace lcpush;
using lcpush::testing::EnvVar;
using lcpush::testing::FakeSubprocess;

TEST_CASE("read stdin until eof sentinel") {
    std::istringstream stream("line one\nline two\nEOF\nignored\n");
    CHECK(editor::read_stdin(stream) == "line one\nline two\n");
}

TEST_CASE("read stdin until end of stream") {
    std::istringstream stream("a\nb\n");
    CHECK(editor::read_stdin(stream) == "a\nb\n");
}

TEST_CASE("editor command prefers visual") {
    EnvVar visual("VISUAL", "code --wait");
    EnvVar editor_var("EDITOR", "nano");
    CHECK(editor::editor_command() == std::vector<std::string>{"code", "--wait"});
}

TEST_CASE("editor command falls back to vi") {
    EnvVar visual("VISUAL", std::nullopt);
    EnvVar editor_var("EDITOR", std::nullopt);
    CHECK(editor::editor_command() == std::vector<std::string>{"vi"});
}

TEST_CASE("strip header removes the instruction line") {
    std::string body = std::string(editor::kSolutionHeader) + "print(1)\n";
    auto stripped = editor::strip_header(body, editor::kSolutionHeader);
    CHECK(stripped.find("print(1)") != std::string::npos);
    CHECK(stripped.find("Paste your solution") == std::string::npos);
}

TEST_CASE("open editor returns none when unchanged") {
    EnvVar editor_var("EDITOR", "true");
    FakeSubprocess proc;
    proc.on_run_interactive = [](const std::vector<std::string>&) {
        return std::optional<int>{0};
    };
    util::SubprocessOverride subprocess_override(&proc);
    CHECK_FALSE(editor::open_editor().has_value());
}

TEST_CASE("open editor returns written content") {
    EnvVar editor_var("EDITOR", "fake");
    FakeSubprocess proc;
    proc.on_run_interactive = [](const std::vector<std::string>& command) {
        std::ofstream out(command.back());
        out << editor::kSolutionHeader << "class Solution: pass\n";
        return std::optional<int>{0};
    };
    util::SubprocessOverride subprocess_override(&proc);
    auto result = editor::open_editor();
    REQUIRE(result.has_value());
    CHECK(result->find("class Solution: pass") != std::string::npos);
    CHECK(result->find("Paste your solution") == std::string::npos);
}

TEST_CASE("open editor returns none for empty buffer") {
    EnvVar editor_var("EDITOR", "fake");
    FakeSubprocess proc;
    proc.on_run_interactive = [](const std::vector<std::string>& command) {
        std::ofstream out(command.back());
        out << "\n\n";
        return std::optional<int>{0};
    };
    util::SubprocessOverride subprocess_override(&proc);
    CHECK_FALSE(editor::open_editor().has_value());
}

TEST_CASE("clipboard command on macos") {
    FakeSubprocess proc;
    proc.on_which = [](const std::string&) {
        return std::optional<std::string>{"/usr/bin/pbpaste"};
    };
    util::SubprocessOverride subprocess_override(&proc);
    auto command = clipboard::clipboard_command("darwin");
    REQUIRE(command.has_value());
    CHECK(*command == std::vector<std::string>{"pbpaste"});
}

TEST_CASE("clipboard command missing tool") {
    EnvVar wayland("WAYLAND_DISPLAY", std::nullopt);
    FakeSubprocess proc;
    util::SubprocessOverride subprocess_override(&proc);
    CHECK_FALSE(clipboard::clipboard_command("linux").has_value());
}

TEST_CASE("clipboard prefers wayland") {
    EnvVar wayland("WAYLAND_DISPLAY", "wayland-0");
    FakeSubprocess proc;
    proc.on_which = [](const std::string& name) {
        return std::optional<std::string>{"/usr/bin/" + name};
    };
    util::SubprocessOverride subprocess_override(&proc);
    auto command = clipboard::clipboard_command("linux");
    REQUIRE(command.has_value());
    CHECK((*command)[0] == "wl-paste");
}

TEST_CASE("clipboard read returns none when whitespace") {
    FakeSubprocess proc;
    proc.on_which = [](const std::string& name) {
        return std::optional<std::string>{"/usr/bin/" + name};
    };
    proc.on_run_capture = [](const std::vector<std::string>&) {
        return util::RunResult{0, "   \n", true};
    };
    util::SubprocessOverride subprocess_override(&proc);
    CHECK_FALSE(clipboard::read().has_value());
}

TEST_CASE("clipboard read returns content") {
    FakeSubprocess proc;
    proc.on_which = [](const std::string& name) {
        return std::optional<std::string>{"/usr/bin/" + name};
    };
    proc.on_run_capture = [](const std::vector<std::string>&) {
        return util::RunResult{0, "class Solution: pass\n", true};
    };
    util::SubprocessOverride subprocess_override(&proc);
    auto content = clipboard::read();
    REQUIRE(content.has_value());
    CHECK(*content == "class Solution: pass\n");
}

TEST_CASE("clipboard read survives a failing tool") {
    FakeSubprocess proc;
    proc.on_which = [](const std::string& name) {
        return std::optional<std::string>{"/usr/bin/" + name};
    };
    proc.on_run_capture = [](const std::vector<std::string>&) {
        return util::RunResult{};  // spawn failure
    };
    util::SubprocessOverride subprocess_override(&proc);
    CHECK_FALSE(clipboard::read().has_value());
}
